source/common/network/happy_eyeballs_connection_impl.cc

```diff
@@ -379,6 +380,41 @@ void HappyEyeballsConnectionImpl::dumpState(std::ostream& os, int indent_level)
   }
 }
 
+std::vector<Address::InstanceConstSharedPtr>
+HappyEyeballsConnectionImpl::sortAddresses(const std::vector<Address::InstanceConstSharedPtr>& in) {
+  std::vector<Address::InstanceConstSharedPtr> address_list = in;
+  for (size_t current = 1; current < address_list.size(); ++current) {
+    // If the current address has a different family than the previous address then
+    // it is in the correct position.
+    if (address_list[current]->type() != Address::Type::Ip ||
+        address_list[current - 1]->type() != Address::Type::Ip ||
+        address_list[current]->ip()->version() != address_list[current - 1]->ip()->version()) {
+      continue;
+    }
+    // Find the first address with a different address family than the current address.
+    size_t next = current + 1;
+    while (next < address_list.size()) {
+      if (address_list[next]->type() != Address::Type::Ip ||
+          address_list[current]->ip()->version() != address_list[next]->ip()->version()) {
+        break;
+      }
+      ++next;
+    }
+    // There are no more addresses with different families so sorting is finished.
+    if (next >= address_list.size()) {
+      break;
+    }
+
+    // Bubble the address up to the current position.
+    for (size_t i = next; i > current; i--) {
```

Comment:
can this also be implemented with [std::rotate](https://en.cppreference.com/w/cpp/algorithm/rotate)?

Reply:
I think the answer is yes, from the std::rotate docs. If I'm reading them correctly, I want to do a "right rotate" not the default "left rotate". I think this means that I will need reverse iterators. I think std::make_reverse_iterator should be able to take a normal iterator to an element and turn it into a reverse iterator to that element. But I've spent an embarrassingly long time trying to make this work and I don't think I'm any closer to being able to do so than when I started :(

I think I want something like
```
  while (current != address_list.end()) {
    auto it = std::find_if(current, address_list.end(),
                           [&](const auto& val){ return !hasSameAddressFamily(*current, val); } );
    if (it == address_list.end()) {
      break;
    }
    auto start = std::make_reverse_iterator(it);
    auto end = std::make_reverse_iterator(current+1);
    std::rotate(start, start + 1, end);
    current++;
  }
```
But this doesn't seem to do what I expect. I think maybe make_reverse_iterator isn't doing what I expect but this is all quite opaque. (if I print the value of, say, "it" and then the value of "start", they don't seem to point to the same element?) I hate to push this back to you, but I'm rather stuck here. Does this code look to you like it *should* be working? Do you have any advice?

Reply:
What you have is what I would expect the solution to be. 

let me give it a try with your branch. you wrote the tests so I can see if I can make it work.

Reply:
OH! I think I finally figured it out. The reverse iterators start one element off from where the forward iterators are pointing to. All my tests pass now. What do you think?

Reply:
You beat me to it. I was also stuck with this one, very easy to rabbit hole with all the pointer arithmetics and the reverse order. These reverse iterators are really unintuitive. 

I think the gotcha is here, from c++ [reverse_iterators](https://en.cppreference.com/w/cpp/iterator/reverse_iterator)

> For a reverse iterator r constructed from an iterator i, the relationship `&*r == &*(i-1)` is always true (as long as r is dereferenceable); thus a reverse iterator constructed from a one-past-the-end iterator dereferences to the last element in a sequence.


Also this implementation is even shorter because we don't need the if case to check if the current IP is equal to the previous.

Reply:
AH! Thanks. That explains it! I didn't see that info on [make_reverse_iterator](https://en.cppreference.com/w/cpp/iterator/make_reverse_iterator), but it's clear as day on that page. Thanks for the pointer!